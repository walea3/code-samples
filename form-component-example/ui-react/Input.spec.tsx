import { mountWithNEL, renderWithNEL } from '@nel-ui/tooling-react/test-utils';
import React from 'react';

import { ReactWrapper } from 'enzyme';
import { Input } from '.';

export const findDataRef = (node: ReactWrapper, targetRef: string): ReactWrapper =>
  node.find(`[data-ref="${targetRef}"]`);

describe('<Input />', () => {
  const requiredProps = {
    id: 'inputId',
  };

  const prefix = 'prefix';
  const suffix = 'suffix';

  const mountComponent = (props) => mountWithNEL(<Input {...requiredProps} {...props} />);

  const renderComponent = (props) =>
    renderWithNEL(<Input {...requiredProps} {...props} />);

  describe('@renders', () => {
    test.each`
      name                                              | props
      ${'default'}                                      | ${{}}
      ${'as inline'}                                    | ${{ inline: true }}
      ${'when disabled'}                                | ${{ disabled: true }}
      ${'with size'}                                    | ${{ size: 5 }}
      ${'with minimum size'}                            | ${{ size: 2 }}
      ${'with below minimum size'}                      | ${{ size: 1 }}
      ${'with size with style'}                         | ${{ size: 1, style: { opacity: 0.5 } }}
      ${'with style'}                                   | ${{ style: { opacity: 0.5 } }}
      ${'with class name'}                              | ${{ className: 'test-input' }}
      ${'with warning label'}                           | ${{ label: 'Input label' }}
      ${'with warning label + className'}               | ${{ label: 'Input label', className: 'custom-input-labelContainer' }}
      ${'with warning label + as inline'}               | ${{ label: 'Input label', inline: true }}
      ${'with warning alert'}                           | ${{ alert: 'warning' }}
      ${'when disabled + with warning alert'}           | ${{ disabled: true, alert: 'warning' }}
      ${'with error alert'}                             | ${{ alert: 'error' }}
      ${'when disabled + with error alert'}             | ${{ disabled: true, alert: 'error' }}
      ${'with prefix'}                                  | ${{ prefix }}
      ${'with suffix'}                                  | ${{ suffix }}
      ${'with prefix & suffix'}                         | ${{ prefix, suffix }}
      ${'with data-ref'}                                | ${{ 'data-ref': 'uid' }}
      ${'with data-ref with label'}                     | ${{ 'data-ref': 'uid', label: 'I am a test :)' }}
      ${'with data-ref with prefix & suffix'}           | ${{ 'data-ref': 'uid', prefix, suffix }}
      ${'with aria-labelledby, label, prefix & suffix'} | ${{ 'aria-labelledby': 'extraInputLabel', label: 'I am a test :)', prefix, suffix }}
      ${'with other props'}                             | ${{ 'data-custom': 'uid' }}
    `('$name', ({ props }) => {
      expect(renderComponent(props)).toMatchSnapshot();
    });

    describe('when focussed', () => {
      test.each`
        name                        | dataRef          | focusEvent | props
        ${'by focussing on input'}  | ${'input'}       | ${'focus'} | ${{}}
        ${'by focussing on prefix'} | ${'inputPrefix'} | ${'focus'} | ${{ prefix }}
        ${'by focussing on suffix'} | ${'inputSuffix'} | ${'focus'} | ${{ suffix }}
      `('$name', ({ dataRef, focusEvent, props }) => {
        const { root, wrapper } = mountComponent(props);
        const el = findDataRef(wrapper, dataRef);

        el.simulate(focusEvent);
        expect(wrapper.render()).toMatchSnapshot();

        el.simulate('blur');
        expect(wrapper.render()).toMatchSnapshot('then blurred');

        root.unmount();
      });
    });

    describe('focusses input', () => {
      test.each`
        name                       | dataRef          | focusEvent | props
        ${'by clicking on prefix'} | ${'inputPrefix'} | ${'click'} | ${{ prefix }}
        ${'by clicking on suffix'} | ${'inputSuffix'} | ${'click'} | ${{ suffix }}
      `('$name', ({ dataRef, focusEvent, props }) => {
        const { root, wrapper } = mountComponent(props);
        const el = findDataRef(wrapper, dataRef);

        el.simulate(focusEvent);
        expect(wrapper.render()).toMatchSnapshot();

        root.unmount();
      });
    });
  });

  describe('@ref', () => {
    test('current is assigned input node', () => {
      const myRef = React.createRef<HTMLInputElement>();

      const { root, wrapper } = mountComponent({ id: 'ref', ref: myRef });
      expect(myRef.current).toEqual(wrapper.find('input').getDOMNode());
      root.unmount();
    });

    test('called with input node', () => {
      const myRef = jest.fn();

      const { root, wrapper } = mountComponent({ ref: myRef });
      expect(myRef).toHaveBeenCalledWith(findDataRef(wrapper, 'input').getDOMNode());
      root.unmount();
    });
  });

  describe('@events', () => {
    test.each`
      name                     | dataRef          | event      | handler
      ${'focussing on input'}  | ${'input'}       | ${'focus'} | ${'onFocus'}
      ${'focussing on prefix'} | ${'inputPrefix'} | ${'focus'} | ${'onFocus'}
      ${'focussing on suffix'} | ${'inputSuffix'} | ${'focus'} | ${'onFocus'}
      ${'blurring input'}      | ${'input'}       | ${'blur'}  | ${'onBlur'}
      ${'blurring prefix'}     | ${'inputPrefix'} | ${'blur'}  | ${'onBlur'}
      ${'blurring suffix'}     | ${'inputSuffix'} | ${'blur'}  | ${'onBlur'}
    `('$name calls event handler', ({ dataRef, event, handler }) => {
      const expectedEvent = {
        // Note: This will return onFocus/onBlur
        [handler]: jest.fn(),
      };

      const { root, wrapper } = mountComponent({
        ...expectedEvent,
        id: 'prefixSuffixEvents',
        prefix,
        suffix,
      });

      findDataRef(wrapper, dataRef).simulate(event);
      expect(expectedEvent[handler]).toHaveBeenCalledTimes(1);
      expect(expectedEvent[handler]).toHaveBeenCalledWith(
        expect.objectContaining({ type: event })
      );

      root.unmount();
    });
  });
});
