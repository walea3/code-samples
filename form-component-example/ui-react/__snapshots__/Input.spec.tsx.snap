// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> @renders as inline 1`] = `
<div
  class="nel-input nel-inlineInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders default 1`] = `
<div
  class="nel-input"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders focusses input by clicking on prefix 1`] = `
<div
  class="nel-input"
  data-ref="inputContainer"
>
  <div
    class="nel-inputPrefix nel-inputPrefixSuffix"
    data-ref="inputPrefix"
    id="inputIdPrefix"
  >
    prefix
  </div>
  <input
    aria-invalid="false"
    aria-labelledby="inputId inputIdPrefix"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders focusses input by clicking on suffix 1`] = `
<div
  class="nel-input"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    aria-labelledby="inputId  inputIdSuffix"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
  <div
    class="nel-inputSuffix nel-inputPrefixSuffix"
    data-ref="inputSuffix"
    id="inputIdSuffix"
  >
    suffix
  </div>
</div>
`;

exports[`<Input /> @renders when disabled + with error alert 1`] = `
<div
  class="nel-input nel-disabledInput nel-errorInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="true"
    class="nel-inputControl"
    data-ref="input"
    disabled=""
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders when disabled + with warning alert 1`] = `
<div
  class="nel-input nel-disabledInput nel-warningInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="true"
    class="nel-inputControl"
    data-ref="input"
    disabled=""
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders when disabled 1`] = `
<div
  class="nel-input nel-disabledInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="input"
    disabled=""
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders when focussed by focussing on input 1`] = `
<div
  class="nel-input nel-focussedInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders when focussed by focussing on input: then blurred 1`] = `
<div
  class="nel-input"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders when focussed by focussing on prefix 1`] = `
<div
  class="nel-input nel-focussedInput"
  data-ref="inputContainer"
>
  <div
    class="nel-inputPrefix nel-inputPrefixSuffix"
    data-ref="inputPrefix"
    id="inputIdPrefix"
  >
    prefix
  </div>
  <input
    aria-invalid="false"
    aria-labelledby="inputId inputIdPrefix"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders when focussed by focussing on prefix: then blurred 1`] = `
<div
  class="nel-input"
  data-ref="inputContainer"
>
  <div
    class="nel-inputPrefix nel-inputPrefixSuffix"
    data-ref="inputPrefix"
    id="inputIdPrefix"
  >
    prefix
  </div>
  <input
    aria-invalid="false"
    aria-labelledby="inputId inputIdPrefix"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders when focussed by focussing on suffix 1`] = `
<div
  class="nel-input nel-focussedInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    aria-labelledby="inputId  inputIdSuffix"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
  <div
    class="nel-inputSuffix nel-inputPrefixSuffix"
    data-ref="inputSuffix"
    id="inputIdSuffix"
  >
    suffix
  </div>
</div>
`;

exports[`<Input /> @renders when focussed by focussing on suffix: then blurred 1`] = `
<div
  class="nel-input"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    aria-labelledby="inputId  inputIdSuffix"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
  <div
    class="nel-inputSuffix nel-inputPrefixSuffix"
    data-ref="inputSuffix"
    id="inputIdSuffix"
  >
    suffix
  </div>
</div>
`;

exports[`<Input /> @renders with aria-labelledby, label, prefix & suffix 1`] = `
<div
  class="nel-fieldControlLabelContainer"
  data-ref="fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="fieldControlLabel"
    for="inputId"
    id="inputIdLabel"
  >
    I am a test :)
  </label>
  <div
    class="nel-input"
    data-ref="inputContainer"
  >
    <div
      class="nel-inputPrefix nel-inputPrefixSuffix"
      data-ref="inputPrefix"
      id="inputIdPrefix"
    >
      prefix
    </div>
    <input
      aria-invalid="false"
      aria-labelledby="extraInputLabel inputIdLabel inputId inputIdPrefix inputIdSuffix"
      class="nel-inputControl"
      data-ref="input"
      id="inputId"
    />
    <div
      class="nel-inputSuffix nel-inputPrefixSuffix"
      data-ref="inputSuffix"
      id="inputIdSuffix"
    >
      suffix
    </div>
  </div>
</div>
`;

exports[`<Input /> @renders with below minimum size 1`] = `
<div
  class="nel-input nel-inlineInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
    size="1"
    style="width:56px"
  />
</div>
`;

exports[`<Input /> @renders with class name 1`] = `
<div
  class="nel-input test-input"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders with data-ref 1`] = `
<div
  class="nel-input"
  data-ref="uid.inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="uid.input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders with data-ref with label 1`] = `
<div
  class="nel-fieldControlLabelContainer"
  data-ref="uid.fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="uid.fieldControlLabel"
    for="inputId"
    id="inputIdLabel"
  >
    I am a test :)
  </label>
  <div
    class="nel-input"
    data-ref="uid.inputContainer"
  >
    <input
      aria-invalid="false"
      class="nel-inputControl"
      data-ref="uid.input"
      id="inputId"
    />
  </div>
</div>
`;

exports[`<Input /> @renders with data-ref with prefix & suffix 1`] = `
<div
  class="nel-input"
  data-ref="uid.inputContainer"
>
  <div
    class="nel-inputPrefix nel-inputPrefixSuffix"
    data-ref="uid.inputPrefix"
    id="inputIdPrefix"
  >
    prefix
  </div>
  <input
    aria-invalid="false"
    aria-labelledby="inputId inputIdPrefix inputIdSuffix"
    class="nel-inputControl"
    data-ref="uid.input"
    id="inputId"
  />
  <div
    class="nel-inputSuffix nel-inputPrefixSuffix"
    data-ref="uid.inputSuffix"
    id="inputIdSuffix"
  >
    suffix
  </div>
</div>
`;

exports[`<Input /> @renders with error alert 1`] = `
<div
  class="nel-input nel-errorInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="true"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders with minimum size 1`] = `
<div
  class="nel-input nel-inlineInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
    size="2"
    style="width:56px"
  />
</div>
`;

exports[`<Input /> @renders with other props 1`] = `
<div
  class="nel-input"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-custom="uid"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders with prefix & suffix 1`] = `
<div
  class="nel-input"
  data-ref="inputContainer"
>
  <div
    class="nel-inputPrefix nel-inputPrefixSuffix"
    data-ref="inputPrefix"
    id="inputIdPrefix"
  >
    prefix
  </div>
  <input
    aria-invalid="false"
    aria-labelledby="inputId inputIdPrefix inputIdSuffix"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
  <div
    class="nel-inputSuffix nel-inputPrefixSuffix"
    data-ref="inputSuffix"
    id="inputIdSuffix"
  >
    suffix
  </div>
</div>
`;

exports[`<Input /> @renders with prefix 1`] = `
<div
  class="nel-input"
  data-ref="inputContainer"
>
  <div
    class="nel-inputPrefix nel-inputPrefixSuffix"
    data-ref="inputPrefix"
    id="inputIdPrefix"
  >
    prefix
  </div>
  <input
    aria-invalid="false"
    aria-labelledby="inputId inputIdPrefix"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders with size 1`] = `
<div
  class="nel-input nel-inlineInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
    size="5"
    style="width:92px"
  />
</div>
`;

exports[`<Input /> @renders with size with style 1`] = `
<div
  class="nel-input nel-inlineInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
    size="1"
    style="opacity:0.5;width:56px"
  />
</div>
`;

exports[`<Input /> @renders with style 1`] = `
<div
  class="nel-input"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
    style="opacity:0.5"
  />
</div>
`;

exports[`<Input /> @renders with suffix 1`] = `
<div
  class="nel-input"
  data-ref="inputContainer"
>
  <input
    aria-invalid="false"
    aria-labelledby="inputId  inputIdSuffix"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
  <div
    class="nel-inputSuffix nel-inputPrefixSuffix"
    data-ref="inputSuffix"
    id="inputIdSuffix"
  >
    suffix
  </div>
</div>
`;

exports[`<Input /> @renders with warning alert 1`] = `
<div
  class="nel-input nel-warningInput"
  data-ref="inputContainer"
>
  <input
    aria-invalid="true"
    class="nel-inputControl"
    data-ref="input"
    id="inputId"
  />
</div>
`;

exports[`<Input /> @renders with warning label + as inline 1`] = `
<div
  class="nel-fieldControlLabelContainer nel-inlineContainer"
  data-ref="fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="fieldControlLabel"
    for="inputId"
    id="inputIdLabel"
  >
    Input label
  </label>
  <div
    class="nel-input"
    data-ref="inputContainer"
  >
    <input
      aria-invalid="false"
      class="nel-inputControl"
      data-ref="input"
      id="inputId"
    />
  </div>
</div>
`;

exports[`<Input /> @renders with warning label + className 1`] = `
<div
  class="custom-input-labelContainer nel-fieldControlLabelContainer"
  data-ref="fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="fieldControlLabel"
    for="inputId"
    id="inputIdLabel"
  >
    Input label
  </label>
  <div
    class="nel-input"
    data-ref="inputContainer"
  >
    <input
      aria-invalid="false"
      class="nel-inputControl"
      data-ref="input"
      id="inputId"
    />
  </div>
</div>
`;

exports[`<Input /> @renders with warning label 1`] = `
<div
  class="nel-fieldControlLabelContainer"
  data-ref="fieldControlLabelContainer"
>
  <label
    class="nel-inputLabel"
    data-ref="fieldControlLabel"
    for="inputId"
    id="inputIdLabel"
  >
    Input label
  </label>
  <div
    class="nel-input"
    data-ref="inputContainer"
  >
    <input
      aria-invalid="false"
      class="nel-inputControl"
      data-ref="input"
      id="inputId"
    />
  </div>
</div>
`;
